include/circt/Dialect/MSFT/MSFTConstructs.td

```diff
@@ -67,3 +67,48 @@ def ChannelOp: MSFTOp<"constructs.channel",
     $input $clk $sym_name `(` $defaultStages `)` attr-dict `:` type($input)
   }];
 }
+
+// Linear, pipelineable datapath.
+def LinearOp : MSFTOp<"hlc.linear", [
+    SingleBlockImplicitTerminator<"OutputOp">,
+    IsolatedFromAbove
+  ]> {
+  let summary = "Model of a linear datapath which can be arbitrarily pipelined";
+  let description = [{
+    Defines a feed-forward datapath which can be scheduled into a pipeline.
+    Due to the feed-forwardness, the inner region is NOT a graph region.
+    Internally, only combinational operations (`comb`, `msft`, `hw`) are allowed.
+    The operation expects exactly one i1-typed port to be tagged with a
+    `msft.clock` attribute.
+
+    Example:
+    ```mlir
+    msft.module @foo(%in0 : i32, %in1 : i32, %in2 : i32, %clk : i1) -> (out: i32) -> {
+      %0 = msft.hlc.linear(%a = %in0, %b = %in1, %c = %in2) clock %clk (i32, i32, i32) -> (i32) {
+        %0 = comb.mul %a, %b : i32
+        %1 = comb.add %0, %c : i32
+        msft.output %1 : i32
+      }
+    }
+    ```
+  }];
+
+  let arguments = (ins Variadic<AnyType>:$ins, I1:$clock);
+  let results = (outs Variadic<AnyType>:$outs);
+  let regions = (region SizedRegion<1>:$datapath);
+
+  let skipDefaultBuilders = 1;
+  let builders = [
+    OpBuilder<(ins "StringAttr":$name, "hw::ModulePortInfo":$ports)>
+  ];
+
+  let extraClassDeclaration = [{
+    Block *getBodyBlock() { return &datapath().front(); }
+
+    // Returns the functional type of this operation.
+    FunctionType getType();
+  }];
+
+  let hasVerifier = 1;
+  let hasCustomAssemblyFormat = 1;
```

Comment:
Can you use asm with `custom`. This looks mostly supported by the asm format. Also, whatever we come up with for the "parent block value use" may be useful in future ops.

Reply:
Just tried with a `custom` parser - first endeavour into this, and i really like it!
``` custom<LinearIns>($ins, type($ins)) `clock` $clock attr-dict `:` type($outs) $datapath}```
with
```mlir
  %0 = msft.hlc.linear(%arg0 : i32 = %in0, %arg1 : i32 = %in1, %arg2 : i32 = %in2) clock %clk : i32 {
    ...
```
However, there doesn't seem to be a way around having a full custom parser for this op. The region argument identifiers parsed during the input list needs to be used when parsing the `$datapath` region.  
If we don't care about this additional syntax (but still want an `IsolatedFromAbove` op) we could just enforce a block within the inner region. Personally i think this looks bad, but accessing and iterating the region through code is exactly the same, so it's purely visual.

```mlir
  %0 = msft.hlc.linear(%in1, %in2, %in3) clock %clk : (i32, i32, i32) {
    ^bb0(%arg0 : i32, %arg1 : i32, %arg2 : i32):
     ...
```